apps/emqx_durable_storage/src/emqx_ds_snapshot_manager.erl

```diff
@@ -0,0 +1,428 @@
+    State0#{rs := RS}.
+
+-spec handle_read_chunk(state(), _Size :: non_neg_integer()) ->
+    {{ok, binary(), next | last} | {error, _Reason :: term()}, state()}.
+handle_read_chunk(#{rs := RS0 = #rs{phase = machine_state, state = MachineState}} = State0, _Size) ->
+    Chunk = term_to_binary(MachineState),
+    RS = RS0#rs{phase = reader_caller_proc},
+    State = State0#{rs := RS},
+    Reply = {ok, Chunk, next},
+    {Reply, State};
+handle_read_chunk(
+    #{rs := RS0 = #rs{phase = reader_caller_proc, reader_caller_proc = {Pid, _}}} = State0, _Size
+) ->
+    Chunk = term_to_binary(Pid),
+    RS = RS0#rs{phase = storage_snapshot},
+    State = State0#{rs := RS},
+    Reply = {ok, Chunk, next},
+    {Reply, State};
+handle_read_chunk(
+    #{rs := RS0 = #rs{phase = storage_snapshot, reader = SnapReader0}} = State0, Size
+) ->
+    case emqx_ds_storage_snapshot:read_chunk(SnapReader0, Size) of
+        {next, Chunk, SnapReader} ->
+            RS = RS0#rs{reader = SnapReader},
+            State = State0#{rs := RS},
+            Reply = {ok, Chunk, next},
+            {Reply, State};
+        {last, Chunk, SnapReader} ->
+            %% TODO: idempotence?
+            ?tp(dsrepl_snapshot_read_complete, #{reader => SnapReader}),
+            _ = complete_read(RS0#rs{reader = SnapReader}),
+            State = State0#{rs := undefined},
+            Reply = {ok, Chunk, last},
+            {Reply, State};
+        {error, Reason} ->
+            ?tp(dsrepl_snapshot_read_error, #{reason => Reason, reader => SnapReader0}),
+            _ = emqx_ds_storage_snapshot:release_reader(SnapReader0),
+            State = State0#{rs := undefined},
+            Reply = {error, Reason},
+            {Reply, State}
+    end.
+
+complete_read(
+    RS = #rs{reader = SnapReader, started_at = StartedAt, reader_caller_proc = {_Pid, MRef}}
+) ->
+    _ = emqx_ds_storage_snapshot:release_reader(SnapReader),
+    demonitor(MRef, [flush]),
+    logger:info(#{
+        msg => "dsrepl_snapshot_read_complete",
+        shard => shard_id(RS),
+        duration_ms => erlang:monotonic_time(millisecond) - StartedAt,
+        read_bytes => emqx_ds_storage_snapshot:reader_info(bytes_read, SnapReader)
+    }).
+
+-spec handle_reader_down(state()) -> state().
+handle_reader_down(State0) ->
+    logger:notice(#{msg => "Reader process down"}),
+    ?tp(dsrepl_reader_down, #{}),
+    cleanup_reader(State0).
+
+cleanup_reader(#{rs := #rs{reader = SnapReader, reader_caller_proc = {_Pid, MRef}}} = State0) ->
+    _ = emqx_ds_storage_snapshot:release_reader(SnapReader),
+    demonitor(MRef, [flush]),
+    State0#{rs := undefined};
+cleanup_reader(State) ->
+    State.
+
+%% Writer (follower state)
+
+-spec handle_accept_chunk(binary(), state()) ->
+    {ok | {error, _Reason :: term()}, state()}.
+handle_accept_chunk(Chunk, #{ws := WS0 = #ws{phase = reader_caller_proc}} = State0) ->
+    ReaderPid = binary_to_term(Chunk),
+    MRef = monitor(process, ReaderPid),
```

Comment:
I'm not really against such approach, but:
1. We must have the guarantee that this DOWN signal won't arrive _before_ last chunk.
2. This has questionable forward-compatibility (probably need to wrap it in some kind of well-defined packet?).

Reply:
> We must have the guarantee that this DOWN signal won't arrive before last chunk.

Great point.

> This has questionable forward-compatibility (probably need to wrap it in some kind of well-defined packet?).

I'm not seeing what makes this chunk more special than the others :see_no_evil: .  You mean serializing the PID is unsafe somehow?